File: src/agent/scorer.py
"""Expected return score calculation."""

import pandas as pd
import numpy as np
from typing import Optional, Dict
from src.utils.indicators import RSI, MACD


class Scorer:
    """Calculate expected return scores for opportunities."""
    
    def __init__(self):
        """Initialize scorer with default parameters."""
        self.rsi_period = 14
        self.rsi_early_entry_min = 40
        self.rsi_early_entry_max = 60
        self.rsi_mid_momentum_min = 60
        self.rsi_mid_momentum_max = 70
        self.rsi_late_momentum_min = 70
        self.rsi_late_momentum_max = 75
        self.rsi_extended_threshold = 75
    
    def score(
        self,
        ticker: str,
        price_data: pd.DataFrame,
        catalyst_strength: float,
        upside_potential: float,
        analyst_target: Optional[float] = None,
        news_sentiment: Optional[float] = None,
    ) -> float:
        """
        Calculate expected return score (0-1).
        
        Formula:
        expected_return_score = 
            (catalyst_strength × 0.40) +
            (momentum_acceleration × 0.30) +
            (upside_potential × 0.20) +
            (timing_factor × 0.10)
        
        Args:
            ticker: Stock ticker symbol
            price_data: DataFrame with columns [Open, High, Low, Close, Volume]
            catalyst_strength: Catalyst strength (0-1)
            upside_potential: Upside potential ratio (0-1), capped at 1.0
            analyst_target: Analyst price target
            news_sentiment: News sentiment score
        
        Returns:
            Expected return score (0-1)
        """
        # Clamp inputs to valid ranges
        catalyst_strength = max(0, min(1.0, catalyst_strength))
        upside_potential = max(0, min(1.0, upside_potential))
        
        # Calculate momentum acceleration
        momentum_acceleration = self._calculate_momentum_acceleration(price_data)
        
        # Calculate timing factor
        timing_factor = self._calculate_timing_factor(price_data)
        
        # Weighted combination
        score = (
            (catalyst_strength * 0.40) +
            (momentum_acceleration * 0.30) +
            (upside_potential * 0.20) +
            (timing_factor * 0.10)
        )
        
        return max(0, min(1.0, score))
    
    def _calculate_momentum_acceleration(self, price_data: pd.DataFrame) -> float:
        """
        Calculate momentum acceleration factor (-1 to +1).
        
        Formula:
        momentum_acceleration = 
            (current_rsi - rsi_5_days_ago) / 50 +
            (macd_histogram_today - macd_histogram_5_days_ago) normalized
        """
        if len(price_data) < 6:
            return 0.0
        
        # RSI momentum
        rsi_values = RSI(price_data["Close"], period=self.rsi_period)
        if len(rsi_values) >= 6:
            rsi_current = rsi_values.iloc[-1]
            rsi_5_days_ago = rsi_values.iloc[-6]
            
            if pd.notna(rsi_current) and pd.notna(rsi_5_days_ago):
                rsi_momentum = (rsi_current - rsi_5_days_ago) / 50
            else:
                rsi_momentum = 0.0
        else:
            rsi_momentum = 0.0
        
        # MACD momentum
        macd_line, signal_line, histogram = MACD(price_data["Close"])
        if len(histogram) >= 6:
            histogram_current = histogram.iloc[-1]
            histogram_5_days_ago = histogram.iloc[-6]
            
            if pd.notna(histogram_current) and pd.notna(histogram_5_days_ago):
                # Normalize MACD change
                macd_momentum = (histogram_current - histogram_5_days_ago) / max(
                    abs(histogram.iloc[-10:]).max(), 0.0001
                )
            else:
                macd_momentum = 0.0
        else:
            macd_momentum = 0.0
        
        momentum_acceleration = (rsi_momentum + macd_momentum) / 2.0
        return max(-1.0, min(1.0, momentum_acceleration))
    
    def _calculate_timing_factor(self, price_data: pd.DataFrame) -> float:
        """
        Calculate timing factor (-0.5 to +0.5).
        
        Timing bonus/penalty:
        +0.5 if entering momentum (RSI 40-60, MACD just crossed bullish)
        +0.25 if mid-momentum (RSI 60-70, MACD positive)
        0 if late momentum (RSI 70-75)
        -0.5 if extended (RSI > 75 or MACD weakening)
        """
        if len(price_data) < 2:
            return 0.0
        
        rsi_values = RSI(price_data["Close"], period=self.rsi_period)
        if len(rsi_values) == 0 or pd.isna(rsi_values.iloc[-1]):
            return 0.0
        
        current_rsi = rsi_values.iloc[-1]
        
        # Check MACD state
        macd_line, signal_line, histogram = MACD(price_data["Close"])
        macd_positive = len(histogram) > 0 and histogram.iloc[-1] > 0
        
        # Early entry (entering momentum)
        if self.rsi_early_entry_min <= current_rsi <= self.rsi_early_entry_max and macd_positive:
            return 0.5
        
        # Mid-momentum
        if self.rsi_mid_momentum_min <= current_rsi <= self.rsi_mid_momentum_max and macd_positive:
            return 0.25
        
        # Late momentum
        if self.rsi_late_momentum_min <= current_rsi <= self.rsi_late_momentum_max:
            return 0.0
        
        # Extended (overbought or MACD weakening)
        if current_rsi > self.rsi_extended_threshold or (len(histogram) > 1 and histogram.iloc[-1] < histogram.iloc[-2]):
            return -0.5
        
        return 0.0
    
    def batch_score(
        self,
        scores_data: Dict[str, Dict],
    ) -> Dict[str, float]:
        """
        Score multiple tickers at once.
        
        Args:
            scores_data: Dict mapping ticker → {
                'price_data': DataFrame,
                'catalyst_strength': float,
                'upside_potential': float,
                'analyst_target': float or None,
            }
        
        Returns:
            Dict mapping ticker → score
        """
        results = {}
        for ticker, data in scores_data.items():
            try:
                score = self.score(
                    ticker=ticker,
                    price_data=data["price_data"],
                    catalyst_strength=data.get("catalyst_strength", 0.0),
                    upside_potential=data.get("upside_potential", 0.0),
                    analyst_target=data.get("analyst_target"),
                    news_sentiment=data.get("news_sentiment"),
                )
                results[ticker] = score
            except Exception as e:
                print(f"Error scoring {ticker}: {e}")
                results[ticker] = 0.0
        
        return results

File: src/agent/scorer.ts
/**
 * Scorer - Expected Return Calculation
 * Links to PRD Section 4 (Scoring System - formal)
 *
 * Formula: expected_return = 0.40 × catalyst + 0.30 × momentum + 0.20 × upside + 0.10 × timing
 */

import { CatalystSignals } from './scanner.js';
import { calculateIndicators, calculateRSI, calculateMACD } from '../utils/indicators.js';
import { CandleData } from '../data/market_data.js';

export interface Score {
  ticker: string;
  expectedReturn: number;
  components: ScoreComponents;
}

export interface ScoreComponents {
  catalystScore: number; // 40% weight - [0,1]
  momentumScore: number; // 30% weight - [-1,1] normalized
  upsideScore: number; // 20% weight - [0,1]
  timingScore: number; // 10% weight - [-0.5,0.5] discrete
}

export class Scorer {
  /**
   * Score a ticker based on PRD multi-factor model
   * Per PRD Section 4: 40% catalyst + 30% momentum + 20% upside + 10% timing
   */
  scoreTickerWithCandles(
    ticker: string,
    candles: CandleData,
    catalyst: CatalystSignals,
    analystTargetPrice?: number
  ): Score {
    if (candles.prices.length < 50) {
      return {
        ticker,
        expectedReturn: 0,
        components: {
          catalystScore: 0,
          momentumScore: 0,
          upsideScore: 0,
          timingScore: 0,
        },
      };
    }

    const currentPrice = candles.prices[candles.prices.length - 1];
    const indicators = calculateIndicators(candles.prices, candles.volumes);

    // Component 1: Catalyst Score (40% weight)
    // PRD: catalyst_strength ∈ [0,1] is sum of triggered signal weights (clipped at 1)
    const catalystScore = Math.min(1.0, Math.max(0.0, catalyst.aggregatedScore));

    // Component 2: Momentum Acceleration (30% weight)
    // PRD: momentum_acceleration ∈ [-1,1] = normalized RSI delta + normalized MACD histogram delta
    const momentumScore = this.calculateMomentumAcceleration(candles);

    // Component 3: Upside Potential (20% weight)
    // PRD: upside_potential ∈ [0,1] = min(1.0, (analyst_target_price - current_price)/current_price)
    const upsideScore = this.calculateUpsidePotential(currentPrice, analystTargetPrice);

    // Component 4: Timing Factor (10% weight)
    // PRD: timing_factor ∈ [-0.5,0.5] with discrete buckets based on RSI/MACD
    const timingScore = this.calculateTimingFactor(indicators);

    // Calculate weighted expected return per PRD formula
    const expectedReturn =
      catalystScore * 0.4 +
      momentumScore * 0.3 +
      upsideScore * 0.2 +
      timingScore * 0.1;

    return {
      ticker,
      expectedReturn: Math.min(1, Math.max(0, expectedReturn)),
      components: {
        catalystScore,
        momentumScore,
        upsideScore,
        timingScore,
      },
    };
  }

  /**
   * Calculate momentum acceleration using RSI and MACD deltas
   * PRD: normalized RSI delta + normalized MACD histogram delta ∈ [-1, 1]
   */
  private calculateMomentumAcceleration(candles: CandleData): number {
    if (candles.prices.length < 30) return 0;

    // Calculate current RSI
    const currentRSI = calculateRSI(candles.prices, 14);

    // Calculate previous RSI (5 periods ago for meaningful delta)
    const prevPrices = candles.prices.slice(0, -5);
    const prevRSI = calculateRSI(prevPrices, 14);

    // RSI delta: range [-100, 100], normalize to [-1, 1]
    const rsiDelta = (currentRSI - prevRSI) / 100;

    // Calculate current and previous MACD histogram
    const currentMACD = calculateMACD(candles.prices);
    const prevMACD = calculateMACD(prevPrices);

    // MACD histogram delta (already in price units, normalize by typical range)
    const macdDelta = currentMACD.histogram - prevMACD.histogram;
    const normalizedMacdDelta = Math.max(-1, Math.min(1, macdDelta / 10)); // Assume ±10 is typical range

    // Combine: normalized RSI delta + normalized MACD delta, bounded to [-1, 1]
    const momentumAcceleration = (rsiDelta + normalizedMacdDelta) / 2;

    return Math.max(-1, Math.min(1, momentumAcceleration));
  }

  /**
   * Calculate upside potential using analyst target price
   * PRD: min(1.0, (analyst_target - current) / current)
   */
  private calculateUpsidePotential(currentPrice: number, analystTargetPrice?: number): number {
    if (!analystTargetPrice || analystTargetPrice <= 0) {
      // No analyst target available - use conservative estimate based on volatility
      return 0.1; // Conservative 10% upside when no target available
    }

    const upside = (analystTargetPrice - currentPrice) / currentPrice;
    return Math.min(1.0, Math.max(0.0, upside));
  }

  /**
   * Calculate timing factor using discrete RSI buckets
   * PRD Section 4:
   * - RSI 40–60 & MACD just crossed bullish: +0.5
   * - RSI 60–70 & MACD positive: +0.25
   * - RSI 70–75: 0
   * - RSI > 75 or MACD weakening: -0.5
   */
  private calculateTimingFactor(indicators: { rsi: number; macdHistogram: number; macd: number; macdSignal: number }): number {
    const rsi = indicators.rsi;
    const macdPositive = indicators.macdHistogram > 0;
    const macdBullish = indicators.macd > indicators.macdSignal;

    // Discrete bucket logic per PRD
    if (rsi >= 40 && rsi <= 60 && macdBullish) {
      return 0.5; // Early entry zone with bullish MACD
    } else if (rsi > 60 && rsi <= 70 && macdPositive) {
      return 0.25; // Momentum continuing
    } else if (rsi > 70 && rsi <= 75) {
      return 0; // Neutral - getting extended
    } else if (rsi > 75 || !macdPositive) {
      return -0.5; // Overbought or weakening
    } else {
      return 0; // Default neutral
    }
  }

  /**
   * Calculate rotation threshold
   * Per PRD Section 6: Threshold = 0.02 for aggressive rotation
   */
  getRotationThreshold(): number {
    return 0.02; // 2% differential triggers rotation
  }
}
